File: daa4.cpp
#include<iostream>
#include<vector>

using namespace std;

int ksackProblem(int n, int m , vector<vector<int>>&items){
      vector<vector<int>>T(n+1,vector<int>(m+1,0));

      for(int i=1;i<=n;i++){
        int profit = items[i-1][0];
        int weight = items[i-1][1];

        for(int j=1;j<=m;j++){
            if(weight <= j){
                T[i][j]= max(T[i-1][j],T[i-1][j-weight]+profit);
            }
            else{
                T[i][j]=T[i-1][j];
            }
        }
      }
      return T[n][m];
}
int main(){
    int n;
    cout<<"Enter the number of Objects : ";
    cin>>n;

    int m;
    cout<<"Enter the knapsack capacity : ";
    cin>>m;

    vector<vector<int>>items(n,vector<int>(2));

    cout<<"Enter the profit and weights for knapsack : ";
    for(int i=0;i<n;i++){
        cin>>items[i][0]>>items[i][1];
    }

    int maxProfit = ksackProblem(n,m,items);
    cout<<"Maximum Profit : "<<maxProfit;

    return 0;
}

// Output - 
// Enter the number of Objects : 4
// Enter the knapsack capacity : 5
// Enter the profit and weights for knapsack : 
// 6 1 
// 5 2
// 3 1
// 8 2
// Maximum Profit : 19

File: daa4.java
import java.util.Scanner;

public class daa4 {

    public static int[][] knapsack(int W, int[] profits, int[] weights, int n) {
        int[][] dp = new int[n + 1][W + 1];

        for (int i = 1; i <= n; i++) {
            for (int w = 1; w <= W; w++) {
                if (weights[i] <= w) {
                    dp[i][w] = Math.max(profits[i] + dp[i - 1][w - weights[i]], dp[i - 1][w]);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }
        return dp;
    }

    public static void printSelectedItems(int[][] dp, int[] weights, int n, int W) {
        int i = n;
        int j = W;

        System.out.println("\nSelected items:");
        while (i > 0) {
            if (j >= weights[i] && dp[i][j] != dp[i - 1][j]) {
                System.out.println("Item " + i + " = 1");
                j = j - weights[i];
            } 
            else {
                System.out.println("Item " + i + " = 0");
            }
            i--;
        }

    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter total number of items: ");
        int n = sc.nextInt();

        int[] weights = new int[n + 1];
        int[] profits = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            System.out.println("Enter weight and profit for item " + i + ": ");
            weights[i] = sc.nextInt();
            profits[i] = sc.nextInt();
        }

        System.out.println("Enter the capacity of the bag: ");
        int W = sc.nextInt();

        int[][] dp = knapsack(W, profits, weights, n);
        int maxProfit = dp[n][W];
        System.out.println("\nMax Profit: " + maxProfit);

        printSelectedItems(dp, weights, n, W);

        sc.close();
    }
}
